src/wg/color/rgb/package.d

```diff
@@ -119,12 +135,178 @@ unittest
+        float b = cast(float)src[2];
+
+        static if (From.ColorSpace.gamma[] != "1")
+        {
+            alias FromGamma = GammaFuncPair!(From.ColorSpace.gamma);
+            r = FromGamma.toLinear(r);
+            g = FromGamma.toLinear(g);
+            b = FromGamma.toLinear(b);
+        }
+        static if (exGamma(From.ColorSpace.id)[] != exGamma(To.ColorSpace.id)[])
+        {
+            import wg.util.math : multiply;
+
+            // TODO: we should do better chromatic adaptation...
+
+            enum toXYZ = From.ColorSpace.rgbToXyzMatrix();
+            enum toRGB = To.ColorSpace.xyzToRgbMatrix();
+            enum mat = multiply(toXYZ, toRGB);
+            float[3] v = multiply(mat, [r, g, b]);
+            r = v[0]; g = v[1]; b = v[2];
+        }
+        static if (To.ColorSpace.gamma[] != "1")
+        {
+            alias ToGamma = GammaFuncPair!(To.ColorSpace.gamma);
+            r = ToGamma.toGamma(r);
+            g = ToGamma.toGamma(g);
+            b = ToGamma.toGamma(b);
+        }
+
+        // convert and return the output
+        static if (To.isOperable)
+        {
+            alias CT = To.ComponentType;
+            return To(cast(CT)r, cast(CT)g, cast(CT)b, cast(CT)src[3]);
+        }
+        else
+        {
+            // each component does something different...
+            static assert(false, "TODO: inoperable (compressed) RGB types");
+        }
+    }
+}
+unittest
+{
+//    // test RGB format conversions
+//    alias UnsignedRGB = RGB!("rgb", ubyte);
+//    alias SignedRGBX = RGB!("rgbx", byte);
+//    alias FloatRGBA = RGB!("rgba", float);
+//
+//    static assert(convertColorImpl!(UnsignedRGB)(SignedRGBX(0x20,0x30,-10)) == UnsignedRGB(0x40,0x60,0));
+//    static assert(convertColorImpl!(UnsignedRGB)(FloatRGBA(1,0.5,0,1)) == UnsignedRGB(0xFF,0x80,0));
+//    static assert(convertColorImpl!(FloatRGBA)(UnsignedRGB(0xFF,0x80,0)) == FloatRGBA(1,float(0x80)/float(0xFF),0,0));
+//    static assert(convertColorImpl!(FloatRGBA)(SignedRGBX(127,-127,-128)) == FloatRGBA(1,-1,-1,0));
+//
+//    static assert(convertColorImpl!(UnsignedRGB)(convertColorImpl!(FloatRGBA)(UnsignedRGB(0xFF,0x80,0))) == UnsignedRGB(0xFF,0x80,0));
+//
+//    // test greyscale conversion
+//    alias UnsignedL = RGB!("l", ubyte);
+//    static assert(cast(UnsignedL)UnsignedRGB(0xFF,0x20,0x40) == UnsignedL(82));
+//
+//    // TODO: we can't test this properly since DMD can't CTFE the '^^' operator! >_<
+//
+//    alias sRGBA = RGB!("rgba", ubyte, false, RGBColorSpace.sRGB);
+//
+//    // test linear conversion
+//    alias lRGBA = RGB!("rgba", ushort, true, RGBColorSpace.sRGB);
+//    assert(convertColorImpl!(lRGBA)(sRGBA(0xFF, 0xFF, 0xFF, 0xFF)) == lRGBA(0xFFFF, 0xFFFF, 0xFFFF, 0xFFFF));
+//
+//    // test gamma conversion
+//    alias gRGBA = RGB!("rgba", byte, false, RGBColorSpace.sRGB_Gamma2_2);
+//    assert(convertColorImpl!(gRGBA)(sRGBA(0xFF, 0x80, 0x01, 0xFF)) == gRGBA(0x7F, 0x3F, 0x03, 0x7F));
```

Comment:
is unittest disabling intended?

Reply:
I just need to rework the test. I reworked most of this code to suit this lib from my other color lib from years ago. We've done a lot of expressions differently here, but principles are all the same.